### ISSUE TITLE
SpriteFont rendering - Can't disable smoothing

### ISSUE THREAD

**Issue by trinith** [MonoGame/MonoGame#6614]
Posted a thread on the community forums for this. Most of the details can be found there :)
http://community.monogame.net/t/spritefont-rendering-cant-seem-to-disable-smoothing/11225?u=trinith

The short of it is, with SamplerState set to PointClamp, fonts still render smoothed and not pixelated. Same deal with setting the SamplerState on the graphics device. I believe the problem might be with whatever turns the SpriteFont file into a texture, because I saved the resulting texture to a file and the smoothing is present there... because of this, it would stand to reason that SamplerState wouldn't matter.

Another community member, Kimimaru, was able to confirm the behaviour that I was seeing. We're wondering if this is a bug.

#### What version of MonoGame does the bug occur on:
- MonoGame 3.7

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

- Windows DX & Android

**Comment by tomspilman** [comment]
The issue is likely that the SpriteFont content processor is smoothing the font when rasterizing it generating a smoothed font.

There was never any option for controlling this in XNA.  Do you know if it generated non-smoothed fonts with the same `.spritefont`?

It could be that FreeType is always smoothing and not obeying some hint in the TTF.

**Comment by tdeeb** [comment]
I looked into this a little since I was having this issue with a font. Though I didn't manage to disable smoothing due to an issue with the font itself, I was able to significantly adjust the output by modifying the `LoadFlags`, `LoadTarget`, and `RenderMode` [when loading the glyph](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/Graphics/Font/SharpFontImporter.cs#L93).

In short, many more settings to adjust fonts are all there and they just needs to be exposed in the pipeline.

**Comment by mrhelmut** [comment]
Being able to control both smoothing and hinting would be great.

**Comment by willmotil** [comment]
> There was never any option for controlling this in XNA.

This maybe true but there was also no pipeline tool in xna.
If the default remains as it is and a option is added then there is no change unless it is explicity made by the user and this is just additional functionality.

> I feel like I was able to do pixelated font rendering in XNA.

Well you can turn on point in the sampler but that doesn't help with anti-aliasing or stretching when the font is rendered to texture on importing. Having the font rasterized without aliasing is just a bit of help it may not entirely solve some artifacts but it would help with the really bad ones.

**Comment by Markvy** [comment]
If anyone needs screenshots from "vintage" XNA, I happen to have it installed, at least for the next week or two when I have to register my copy of Visual C# Express 2010, which I doubt is possible now.

**Comment by tdeeb** [comment]
A comparison with XNA's output would be most welcome!

**Comment by Markvy** [comment]
Well, if you can point me to a spritefont file and some C# code to run, I can paste a screenshot.  (Note that I don't have MonoGame installed.)  I can say that I'm definitely getting font smoothing, but my code doesn't quite match what's in the forum thread.  Alternatively, I could just attach the xnb files that get generated?

**Comment by trinith** [comment]
> Well, if you can point me to a spritefont file and some C# code to run, I can paste a screenshot. (Note that I don't have MonoGame installed.) I can say that I'm definitely getting font smoothing, but my code doesn't quite match what's in the forum thread. Alternatively, I could just attach the xnb files that get generated?

If you check the link at the very beginning, I posted some code in the thread on the MG forums. You can just make the sprite font via Visual Studio.  Any standard font (ie, Consolas) will do. If you need one though, let me know and I can paste the XML :)

**Comment by Markvy** [comment]
That code didn't compile on vintage XNA because that overload of SpriteBatch.Begin() didn't exist.  Once we fix that minor issue...

![image](https://user-images.githubusercontent.com/8607484/81515377-c3f3b300-9301-11ea-87ec-e07a6c07197a.png)

**Comment by ParanoidCactus** [comment]
I spent some time looking into this over the past couple of days.

Changing the rendering mode for glyphs is trivial. A simple matter of changing the RenderMode argument passed into GlyphSlot.RenderGlyph() in SharpFontImporter. But, this will obviously need to be added as an optional property in FontDescription so users can specify in the XML file whether or not they want font smoothing enabled.



That said, when testing this out I discovered a problem with the way pixel data for 1bpp glyphs was being parsed.

Below is an updated to SharpFontImporter's ImportGlyph method that parses the data as per the FreeType specs that can be found here: [https://www.freetype.org/freetype2/docs/reference/ft2-basic_types.html#ft_bitmap](https://www.freetype.org/freetype2/docs/reference/ft2-basic_types.html#ft_bitmap)

With 1bpp glyphs the buffer always contains an even number of bytes per row of pixels regardless of the actual pixel width of the glyph. As such, the buffer may contain padding bytes at the end of each row. The number of bytes per row is the absolute value of the bitmap.Pitch property. If Pitch is negative the rows in the buffer are stored in reverse order, so the row order needs to be flipped when converted. 

I also discovered the BufferData byte array that SharpFont outputs seems to be truncated when glyphs are narrow, such as "i" or "!". Not sure if that's a bug with SharpFont or something else as I couldn't see any obvious issues when I looked at the SharpFont source code. Either way, I've changed it to address the buffer directly with IntPtr.

```
private Glyph ImportGlyph(char character, Face face, bool useSmoothing)
{
	uint glyphIndex = face.GetCharIndex(character);
	face.LoadGlyph(glyphIndex, LoadFlags.Default, LoadTarget.Normal);
	face.Glyph.RenderGlyph(useSmoothing ? RenderMode.Normal : RenderMode.Mono);

	// Render the character.
	BitmapContent glyphBitmap = null;
	if (face.Glyph.Bitmap.Width > 0 && face.Glyph.Bitmap.Rows > 0)
	{
		glyphBitmap = new PixelBitmapContent<byte>(face.Glyph.Bitmap.Width, face.Glyph.Bitmap.Rows);
		byte[] gpixelAlphas = new byte[face.Glyph.Bitmap.Width * face.Glyph.Bitmap.Rows];
		// If the character bitmap has 1bpp we have to expand the buffer data to get the 8bpp pixel data.
		// 1bpp glyphs always have an even number of bytes per row regardless of the width of the glyph, so they may contain unused padding bytes they need to be skipped.
		// The absolute value of bitmap.Pitch is the number of bytes per row.
		// Each non-padding byte in bitmap.Buffer contains the values of up to 8 pixels.
		// If bitmap is of width 10, each row has 2 bytes with 10 valid bits, and the last 6 bits of 2nd byte must be discarded.
		// If bitmap.Pitch is negative the row order of the glyph is reversed (ie. glyph bitmap data is upside down).
		// Note: The reason for using IntPtr to read bitmap.Buffer is because the bitmap.BufferData array may be truncated when a glyph is narrow (bug in SharpFont?)
		if (face.Glyph.Bitmap.PixelMode == PixelMode.Mono)
		{
			// Offset into bitmap.Buffer
			int bufferOffset = 0;
			// rowOffset is the offset in gpixelAlphas to the current row. rowOffsetInc is the amount to increment/decrement rowOffset for each row.
			int rowOffset = face.Glyph.Bitmap.Pitch < 0 ? (face.Glyph.Bitmap.Rows - 1) * face.Glyph.Bitmap.Width : 0, rowOffsetInc = face.Glyph.Bitmap.Pitch / Math.Abs(face.Glyph.Bitmap.Pitch) * face.Glyph.Bitmap.Width;
			for (int i = 0; i < face.Glyph.Bitmap.Rows; i++)
			{
				//get pointer to this row of the bitmap buffer
				IntPtr bufferPtr = IntPtr.Add(face.Glyph.Bitmap.Buffer, bufferOffset);

				// Width in pixels of each row
				int width = face.Glyph.Bitmap.Width, xOffset = 0;
				while(width > 0)
				{
					// Number of bits that contain valid data in the current byte.
					int stride = MathHelper.Min(8, width);
					// Read the next byte from the buffer.
					byte b = Marshal.ReadByte(bufferPtr);
					// Expand each bit to a byte and copy to pixeldata.
					ExpandByteAndCopy(b, stride, gpixelAlphas, rowOffset + xOffset);
					xOffset += stride;
					width -= stride;
					bufferPtr = IntPtr.Add(bufferPtr, 1);
				}
			  // Move to the start of the next row.
			  bufferOffset += Math.Abs(face.Glyph.Bitmap.Pitch);
			  rowOffset += rowOffsetInc;
			}
		}
		else
			Marshal.Copy(face.Glyph.Bitmap.Buffer, gpixelAlphas, 0, gpixelAlphas.Length);

		glyphBitmap.SetPixelData(gpixelAlphas);
	}

	if (glyphBitmap == null) 
	{
		var gHA = face.Glyph.Metrics.HorizontalAdvance >> 6;
		var gVA = face.Size.Metrics.Height >> 6;

		gHA = gHA > 0 ? gHA : gVA;
		gVA = gVA > 0 ? gVA : gHA;

		glyphBitmap = new PixelBitmapContent<byte>(gHA, gVA);
	}

	// not sure about this at all
	var abc = new ABCFloat ();
	abc.A = face.Glyph.Metrics.HorizontalBearingX >> 6;
	abc.B = face.Glyph.Metrics.Width >> 6;
	abc.C = (face.Glyph.Metrics.HorizontalAdvance >> 6) - (abc.A + abc.B);

	// Construct the output Glyph object.
	return new Glyph(character, glyphBitmap)
	{
		XOffset = -(face.Glyph.Advance.X >> 6),
		XAdvance = face.Glyph.Metrics.HorizontalAdvance >> 6,
		YOffset = -(face.Glyph.Metrics.HorizontalBearingY >> 6),
		CharacterWidths = abc
	};
}
```


Further to this, I think it's also valuable to give people the option of choosing a different font face index for their fonts. Some fonts are stored as font collections that contain multiple fonts under the same font family.

![font-sample](https://user-images.githubusercontent.com/52683885/103450718-58d27300-4d0e-11eb-9606-e7bf7b0c4bd2.png)

As an example, the default Japanese font for Windows is MS Gothic, but it contains 3 fonts: MS Gothic, MS UI Gothic and MS PGothic. The above image show font face 0 and font face 1. Font face 0 uses oddly monospaced Latin characters and isn't hinted as well as font face 1, so doesn't look as clean when rendered at this small size.

Of course this would also require another property to be added to FontDescription and another change in SharpFontImporter.cs (although only one line this time).

```
private Face CreateFontFace(FontDescription options, string fontName)
{
	try
	{
		const uint dpi = 96;
		var face = lib.NewFace(fontName, options.FontFaceIndex);
		var fixedSize = ((int)options.Size) << 6;
		face.SetCharSize(0, fixedSize, dpi, dpi);

		if (face.FamilyName == "Microsoft Sans Serif" && options.FontName != "Microsoft Sans Serif")
			throw new PipelineException(string.Format("Font {0} is not installed on this computer.", options.FontName));

		return face;

		// A font substitution must have occurred.
		//throw new Exception(string.Format("Can't find font '{0}'.", options.FontName));
	}
	catch
	{
		throw;
	}
}
```

Anyway, I hope all of this is helpful when you get around to looking in to this issue/feature.
